fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,480 @@
+        if self.log:
+            x = F.log_softmax(x, dim=-1, dtype=torch.float32)
+        else:
+            x = F.softmax(x, dim=-1, dtype=torch.float32)
+        assert x.dtype == torch.float32
+        return x
+
+
+class BaselineSoftmaxNllLoss(BaselineSoftmax):
+    """ Baseline that does an output projection, a softmax & a NLL loss (cross-entropy).
+
+        See BaselineSoftmax above. Constructor is the same. Only difference is in the
+        forward function.
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        """Forward that directly compute the loss."""
+        assert kwargs == {}
+        input, target = input
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        x = super().forward(input, target)
+        x = F.nll_loss(x, target, reduction="sum")
+        return x
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
+
+    @staticmethod
+    def get_max(i: torch.Tensor, w: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _m = torch.matmul(i, w.T)
+        if fp:
+            _m = _m.float()
+        _m = _m.max(dim=1)[0]
+        return _m
+
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any,
+        i: torch.Tensor,
+        w: torch.Tensor,
+        kernel_obj: "MemoryEfficientVocabOutput",
+        w_idx: int,
+        w_split_size: int,
+        split_dim: int,
+    ) -> torch.Tensor:
+        """Forward function that computes the max, without saving activations."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG max fwd")
+        ctx.save_for_backward(i, w)
+        ctx.kernel_obj = kernel_obj
+        ctx.w_idx = w_idx
+        ctx.w_split_size = w_split_size
+        ctx.args = {}
+        assert split_dim == 0
+        with torch.no_grad():
+            return GetMaxFunction.get_max(i, w, kernel_obj.fp_max)
+
+    @staticmethod
+    def backward(ctx: Any, *args: Any) -> Any:
+        """Recompute the forward max and backward grad.
+
+           Accumulate the grad to the right split of the full grad.
+        """
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG max bwd")
+        assert len(args) == 1
+        # Gradients should already exist due to TargetScoreFunction's backward.
+        assert ctx.kernel_obj.proj_weight.grad is not None
+
+        # Get saved i and w.
+        i, w = ctx.saved_tensors
+        assert i.requires_grad
+        assert w.requires_grad
+        i = i.detach().requires_grad_(True)
```

Comment:
the sequence here (detach, FW/BW with autograd, then no_grad again) is pretty confusing without much context. Could you add some explanations ? I'm a bit worried about future maintainability

Reply:
Yes, very true! I will add more comments in the code around this.

Reply:
thanks for adding the comments! they definitely help but I think we need a doc to be honest :) this is pretty complex and context is pretty low with just the code.